multi-element litter, mass checks and diagnostics

rgknox opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description:
<!--- Describe your changes in detail -->
<!--- please add issue number if one exists -->

This changeset refactors the core FATES code to cycle multiple elements beyond carbon12.  This mostly is represented in three pieces of memory, the litter, diagnostics used in writing history diagnostics and mass balance checking and accounting.  These changes have significant impacts on how we process input and output boundary conditions to the model.

This PR should be synchronized with CTSM PR: .


### Collaborators:
<!--- List names of collaborators or people who have interacted -->
<!--- in bringing about this set of changes -->
<!--- consultation, discussions, etc. -->

@ckoven, Bill Riley, @jkshuman, @walkeranthonyp, Fates Modeling Team

### Expectation of Answer Changes:
<!--- Please describe under what conditions, if any, -->
<!--- the model is expected to generated different answers -->
<!--- from the master version of the code -->

The results in this PR should give reasonably similar results to previous simulations. However, so much code has changed, that I do expect round-off errors to be large enough to prevent B4B comparisons from passing (however, they may).

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [ ] I have updated the in-code documentation .AND. (the [technical note](https://github.com/NGEET/fates-docs) .OR. the wiki) accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/master/CONTRIBUTING.md) document.
- [ ] FATES PASS/FAIL regression tests were run
- [ ] If answers were expected to change, evaluation was performed and provided


### Test Results:

TBD

<!--- Non-trivial changes require the PASS/FAIL regression tests. -->
<!--- If changes to code are NOT expected to change answers, tests must -->
<!--- be run against a baseline. -->

CTSM (or) E3SM (specify which) test hash-tag:

CTSM (or) E3SM (specify which) baseline hash-tag:

FATES baseline hash-tag:

Test Output:

<!--- paste in test results here -->


<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

rgknox commented:
tests are coming along:
Here is a 35 year comparison between this branch and master, at BCI, same parameters, no fire, disturbance on, reproduction on and two tropical trees.

[test results - disturbance and reproduction](https://drive.google.com/file/d/1Fu-7c_FOgyoj3zyTJ6qKrYLYp2umcAXT/view?usp=sharing)

Tests Already Passed:
- Prescribed Physiology
- No-recruitment

Next tests (which I will incrementally turn on): 
- inventory initialization
- seed_rain > 0
- ST3
- logging on
- fire on
- frbstore > 0
- 6pfts (trees only)
- 12pfts (w shrubs and grasses)
- F10 grid

ckoven commented:
great! looks like the results are indistinguishable except that the carbon balance error is smaller.

rgknox commented:
Tested with external seed rain and storage to seed (frbstore > 0):
https://drive.google.com/file/d/1nrodAze5FidkWlFL6lgOSlYnFEHNpwox/view?usp=sharing

rgknox commented:
Conducted comparison with master after merging up to tag 1.27.1.    Same test, uses normal features along with seed-rain and storage-to-seed active on two tropical evergreens.  Answers agree with master.  This test also adds 4 new diagnostics, the total litter in fines and cwd, below and above ground.

https://drive.google.com/file/d/1SQvnbVcGLkNMrXU_2arj0LHQ_xVuoQPh/view?usp=sharing

rgknox commented:
tests with logging active ( once per year on 30th day) show same results as base:
https://drive.google.com/file/d/1Bki3MMg044R7Jb3_U73uec7yBcZnJxv3/view?usp=sharing

rgknox commented:
Compared the current set of changes against PR #544 , at BCI, both with spitfire on, two tropical evergreens, default parameter set. Results are showing good agreement:

https://drive.google.com/file/d/1uUTBKh8NvRdkSEeGgvBrUvu4KzMfosOx/view?usp=sharing

rgknox commented:
Here is a comparison with the PR #544  refactors, I'm getting pretty close agreement with this set of changes now:
https://drive.google.com/file/d/1xzVv-2TgGQYX9nyKdi7PSr-0Vs5lb60s/view?usp=sharing

rgknox commented:
Compared multi-element branch with master with ST3.  Note that when dynamics are not on, many of the variables I have been comparing are not active.
https://drive.google.com/file/d/1vm0S9xYuC9CmskI7f9FJbsoEXnCl9buL/view?usp=sharing

rgknox commented:
Comparing inventory initialized BCI runs, with logging and fire on, multi-element vs baseline:
https://drive.google.com/file/d/1LO-iRKOriQjSvZrDmXXO3m4SY-In9hOU/view?usp=sharing

Note: Having fire turned on is a good way to test feedback effects from differential litter generation.

rgknox commented:
The following is a baseline comparison with the 6 default trees in the parameter file, fire and logging are turned on:
https://drive.google.com/file/d/1_g1vca73S7eoyHDEaV8OkmPDNrtK3xlW/view?usp=sharing

I am still interpreting these tests.  The results are similar, but not so simimilar to make an easy assement of preserving base behavior.

Note* The baseline I am using contains adjustments like fixes to fire, found in #544

rgknox commented:
Here is my baseline branch: https://github.com/rgknox/fates/tree/master-1.27.1-mefixes

rgknox commented:
Actually, there is a subtle but important difference that I needed to fix in my base version.  I had previously had to expand the number of allowable cohorts to accomodate space in the restart file for the litter arrays.  But I had recently reduced that space requirement, which thus reduces the cohort requirement.  But, I didn't revert that in my base... retesting.

rgknox commented:
Merged in Greg's PR with module default privitization. I really like the changes.  I did not realize, having public default modules was allowing for some weird stuff to happen. For instance, If I declared a global in FatesLitterMod, but used it in EDTypesMOd;  I could use that variable, globally, in other modules by pointing to the instance in EDTypesMod.  So its like pointing to a copy.  While I believe this gives the right results, it was making the code confusing.  Private by default now prevents this from happening, which is good!

Here is my check to make sure that the merge was correct:
https://drive.google.com/file/d/1bNa5_jN6ytWcJULeofPKmrF37IUfAxC8/view?usp=sharing

I've also continued testing and comparing the multi-element changes with a baseline with more PFTs and have noticed a small but subtle difference in the runs.  

For instance, with no disturbance, and no external seed rain, there is almost no detectable differences between the test and the base:  https://drive.google.com/file/d/1UvilU_q2Or-QZ89jmZLMrlL7DvVRc2N_/view?usp=sharing

When I turn on external seed rain, these tiny differences crop up. Yet, my sense is that the old and new algorithms are mathematically the same, they may have such a different order of operations that results are slightly different: https://drive.google.com/file/d/1DFh3nqP-XtOXGO1qwkYWhw9LV-q9jy2F/view?usp=sharing

I'm still in the process of assessing these small differences, I want to make sure that any changes in the algorithm and results on this multi-element branch are identified.

rgknox commented:
Tests with 6-pfts (all trees), with distrubance, external seed rain, and fire, are now matching baselines (which include PR #544):

https://drive.google.com/file/d/1Nz6ec9QEgQ0cv_KRWZTHcQm3ln8uMara/view?usp=sharing

jkshuman commented on biogeochem/FatesLitterMod.F90:
```diff
@@ -0,0 +1,429 @@
+  ! "litter" means all organic material that is no longer associated with a live plant.
+  ! Also, in FATES we only track "un-fragmented" and "un-decomposed" litter. This
+  ! is a decision of pragmatism, as FATES is not a soil decomposition model, yet FATES
+  ! does need to retain litter for fire calculations.  THerefore, we raitain
```
```suggestion
  ! does need to retain litter for fire calculations.  Therefore, we retain
```

rgknox commented:
Thanks for the feedback @walkeranthonyp.  I think @JunyanDing has been thinking about adding cohort-scale root fractions, but we haven't checked in recently about the design of that code.  Junyan, perhaps it might be a good idea to take a look at the changes in this PR and see if they have any fundamental conflicts or problems with what you want to do?  The intention was to break away from root fractions being stored at the patch scale, and instead it calculates root-depth fractions on-demand, which should be useful when we want to calculate them for each cohort separately.

JunyanDing commented:
Thanks @rgknox  for noticing me about this PR. Yes, we have the same intention. In addition to calculate root-depth fraction for each cohort, I think to make roota, rootb, and effective rooting depth to be calculated in the model instead of as fixed parameters. In FATESHydro, fixed root parameters will cause the roots to grow in places where there is no water.  But, I haven't figure out the best way to do it.

jkshuman commented on biogeochem/EDCohortDynamicsMod.F90:
```diff
@@ -816,7 +768,147 @@ subroutine terminate_cohorts( currentSite, currentPatch, level )
+       flux_diags => csite%flux_diags(el)
+
+       do c=1,ncwd
+                   
```
add a comment flag for the speed readers like myself to slow me down and help me catch ag vs bg
```suggestion

! above ground CWD
```

jkshuman commented on biogeochem/EDCohortDynamicsMod.F90:
```diff
@@ -816,7 +768,147 @@ subroutine terminate_cohorts( currentSite, currentPatch, level )
+          litt%ag_cwd(c) = litt%ag_cwd(c) + plant_dens * &
+               (struct_m+sapw_m)  * SF_val_CWD_frac(c) * &
+               EDPftvarcon_inst%allom_agb_frac(pft)
+       
```
```suggestion

! below ground CWD
```

jkshuman commented on biogeochem/EDCohortDynamicsMod.F90:
```diff
@@ -816,7 +768,147 @@ subroutine terminate_cohorts( currentSite, currentPatch, level )
+                  (1.0_r8 - EDPftvarcon_inst%allom_agb_frac(pft)) * &
+                  csite%rootfrac_scr(sl)
+          enddo
+
```
```suggestion

! above ground
```

jkshuman commented on biogeochem/EDCohortDynamicsMod.F90:
```diff
@@ -816,7 +768,147 @@ subroutine terminate_cohorts( currentSite, currentPatch, level )
+          flux_diags%cwd_ag_input(c)  = flux_diags%cwd_ag_input(c) + &
+                (struct_m+sapw_m) * SF_val_CWD_frac(c) * &
+                EDPftvarcon_inst%allom_agb_frac(pft) * nplant
+
```
```suggestion

! below ground
```

rgknox commented on biogeochem/FatesLitterMod.F90:
```diff
@@ -0,0 +1,429 @@
+                                              ! (twig,s branch,l branch, trunk)
+
+   integer, public, parameter :: ndcmpy = 3   ! number of "decomposability" pools in
+                                              ! fines (lignin, cellulose, labile)
```
@jkshuman ,
All litter, including CWD will ultimately be partitioned into the 3 decomposability pools.  So this comment should be changed.  
I must had put "fines" in there probably because we are only tracking those 3 pools in FATES for the fine tissues (as prognostic variables), whereas for the CWD we partition into those pools only as it is being sent from FATES to the BGC model.

rgknox commented on biogeochem/FatesLitterMod.F90:
```diff
@@ -0,0 +1,429 @@
+
+   integer, public, parameter :: ilabi  = 1   ! Array index for labile portion
+   integer, public, parameter :: icell  = 2   ! Array index for cellulose portion
+   integer, public, parameter :: ilign  = 3   ! Array index for the lignin portion
```
@jkshuman , and labile would also be 2 more... icellul ?  

The neat freak in me wants these to be the same string length.  I suppose I could just close my eyes and write icellulose

jkshuman replied:
I can live with it as is, but updating labile and lignin would increase the readability. The "icell" is easier to self complete in my mind. But if I spent more time with these params, I am sure the short version would be fine.

ckoven replied:
given that not everyone in the fates community is likely totally familiar with decomposition models, I second @jkshuman's request of using the longer forms of these indices.

rgknox commented on biogeochem/EDCohortDynamicsMod.F90:
```diff
@@ -816,7 +768,147 @@ subroutine terminate_cohorts( currentSite, currentPatch, level )
+    ! on a vegetation cohort, into the litter pool.
+    ! 
+    ! Important: (1) This IS NOT turnover, this is not a partial transfer.
+    !            (2) This is from a select number of the cohort
```
ok, will elaborate

rgknox commented on biogeochem/EDCohortDynamicsMod.F90:
```diff
@@ -358,81 +329,110 @@ end subroutine create_cohort
-     ! This is the call to allocate the data structures in the PRT object
-     ! This call will be extended to each specific class.
+       write(fates_log(),*) 'You have not specified the boundary conditions for the'
+       write(fates_log(),*) 'CNP with flexible stoichiometries hypothesis. Please do so. Dude.'
```
ha, that was one of my cheekier coding moments

rgknox replied:
although, "I'm sorry Dave, I'm afraid I can't do that." sounds suitably awesome as well

rgknox commented on biogeochem/EDCohortDynamicsMod.F90:
```diff
@@ -585,9 +582,9 @@ subroutine zero_cohort(cc_p)
     currentcohort%year_net_uptake(:) = 999._r8 ! this needs to be 999, or trimming of new cohorts will break. 
     currentcohort%ts_net_uptake(:)   = 0._r8
-    currentcohort%seed_prod          = 0._r8
     currentcohort%fraction_crown_burned = 0._r8 
```
I'm glad you brought this up @jkshuman , this has been on my mind too.
In my opinion, every variable, upon allocation should be "nan'd".  
 The power of Nan'ing things, is that it makes it clear the variable has not been set since it was defined, which is so useful for debugging and checking the code.  I don't like the idea of doing a blanket "zero'ing" of all variables after allocation.  Once we zero things, its tough to tell at any given time in the call sequence, if a variable is zero because an actual process wants it to be zero, or because we never called a process to modify it.
But there are some things, like variables that are accumulated over several different processes (like mass conservation checks, or fluxes, diagnostics), that I do like to see zero'd as a group.  But, outside of this type of scenario, in my opinion, I think we should avoid blanket zero'ing (particularly prognostic (time-integrated, based on fluxes) variables, unless there is intent, like zero'ing iniital litter mass in a spin-up run).


Performing an audit of all cohort and patch level variables, and addressing their nan'ing and zero'ing has been on my to-do list for a while.

rgknox commented on biogeochem/EDCohortDynamicsMod.F90:
```diff
@@ -648,74 +646,71 @@ subroutine terminate_cohorts( currentSite, currentPatch, level )
     integer :: levcan      ! canopy level
     !----------------------------------------------------------------------
```
good idea, will do

rgknox commented on biogeochem/EDPatchDynamicsMod.F90:
```diff
@@ -78,7 +102,27 @@ module EDPatchDynamicsMod
+
+  real(r8), parameter :: existing_litt_localization = 1.0_r8
+  real(r8), parameter :: treefall_localization = 0.0_r8
+  real(r8), parameter :: burn_localization = 0.0_r8
```
yes, these parameters are just in-code right now, not in the file

rgknox commented on biogeochem/EDPatchDynamicsMod.F90:
```diff
@@ -991,452 +1175,703 @@ subroutine set_patchno( currentSite )
+    real(r8) :: retain_frac          ! the fraction of litter mass retained by the donor patch
+    real(r8) :: bcroot               ! amount of below ground coarse root per cohort kg
+    real(r8) :: bstem                ! amount of above ground stem biomass per cohort kg
     real(r8) :: burned_leaves        ! amount of tissue consumed by fire for leaves. KgC/individual/day
```
yes, this can be removed, nice catch

rgknox commented on biogeochem/EDPhysiologyMod.F90:
```diff
@@ -1189,220 +1576,345 @@ subroutine CWD_Input( currentSite, currentPatch)
-                 (struct_c + sapw_c), dead_n, dead_n_natural, dead_n_ilogging, dead_n_dlogging
-            call endrun(msg=errMsg(sourcefile, __LINE__))
-         endif
+              (leaf_m + fnrt_m + store_m ) * & 
```
This part is hard to tell what is happening unless you open up the code for both old and new, and have them open next to each other.  I saw this particular diff, and I thought the pools going to the resources_management trackers were being changed, but after looking again, its not.  See this line up above:

https://github.com/NGEET/fates/pull/524/files#diff-b2580903c042bbb143bb267a7b882e3fL1309

rgknox replied:
also, note that the units on the number of dead plants has changed, that is why I don't add hlm_freq_day to the calculations anymore, because the death rates are already in indiv/day.

https://github.com/NGEET/fates/pull/524/files#diff-b2580903c042bbb143bb267a7b882e3fR1723

rgknox commented:
@jkshuman , I think I addressed all your comments, and accepted your suggestions.  Let me know if I missed anything.  I went into the EDPhysiology code and double checked how the resource_management diagnostics are filled, and I think the refactors maintain what is in master.

rgknox commented:
I ran some comparison simulations with an updated base branch that is based off of tag 1.29.  Things are looking good.  In this example, both the test and base are started with inventory at BCI, simulate for 8 years, and have logging, fire, disturbance and recruitment all active. I've also done similar tests with prescribed physiology mode as well.  

https://drive.google.com/file/d/1QIROC4-NDLlX4veNpMvsPYAXVk49Qhtj/view?usp=sharing

Will continue testing, next is restart tests (again), hydro and then a long smoke test with the F10 grid. If anyone wants to get reviews in on this, please do so, hopefully we can get this integrated very soon.

glemieux commented on biogeophys/FatesPlantRespPhotosynthMod.F90:
```diff
@@ -313,28 +329,9 @@ real(r8) :: tcwood             ! Temperature response function for wood
 
+
                ! ------------------------------------------------------------------------
                ! Part VI: Loop over all leaf layers.
```
Minor correction: renumber parts [given move](https://github.com/rgknox/fates/commit/3234d4ad23d54a577874e98e06d8420362210fe0#diff-758dbf12610469ab3e6dc7ca96e76e27) of Part V: `set_root_fraction` to above the do loop.

rgknox replied:
nice catch, will do

rgknox commented on biogeochem/FatesLitterMod.F90:
```diff
@@ -0,0 +1,429 @@
+    real(r8),intent(in) :: init_seed
+    real(r8),intent(in) :: init_seed_germ
+    
+    this%ag_cwd(:)              = init_leaf_fines
```
They only initialize variables that describe "states".  ie, the variables that hold the critical instantaneous memory of how much stuff we have. Or in other words, things that are prognosed (prognosticated?), or integrated in time.
Everything else is a flux, a change, or a diagnostic.

rgknox replied:
wait, I see what you mean, yeah, good catch, will fix!

rgknox commented on main/EDPftvarcon.F90:
```diff
@@ -2399,7 +2401,31 @@ subroutine FatesCheckParams(is_master, parteh_mode)
+      integer, intent(in) :: dcmpy
+      real(r8) :: decompy_frac        
+      
+      select case(dcmpy)
```
@glemieux , I couldn't find a reference to `fr_lab`.  
Their are currently only 3 decomposability pools, labile, cellulose and lignins. And in this case, I think we are calling the function for all 3 in sequence.  Did I miss something?

glemieux replied:
`fr_lab` is located [here in EDPftvarcon](https://github.com/rgknox/fates/blob/3d56ddd8b5c00ac4ee98db7aedf9fa34f1097d74/main/EDPftvarcon.F90#L76-L78).  My question came up when I looked at the new `GetDecompyFrac` function and noticed that the function was being called for both [leaf litter](https://github.com/NGEET/fates/blob/acfd8d7603202501ad7d2605b59b570a502d54b0/biogeochem/EDPatchDynamicsMod.F90#L1502-L1508) and [fine root litter](https://github.com/NGEET/fates/blob/acfd8d7603202501ad7d2605b59b570a502d54b0/biogeochem/EDPatchDynamicsMod.F90#L1516-L1525) in EDPatchDynamicsMod, but only the leaf litter values are being used in the function definition.

rgknox replied:
@rgknox  facepalms.  Thanks, @glemieux, I completely missed that we have different decomposability fractions for the two. Dang, that was close, glad you caught that.

glemieux commented on biogeochem/FatesAllometryMod.F90:
```diff
@@ -2056,6 +2058,18 @@ subroutine set_root_fraction(root_fraction, ft, zi, lowerb, icontext )
        call endrun(msg=errMsg(sourcefile, __LINE__))
     end select
 
+!    if( abs(sum(root_fraction)-1.0_r8) > 1.e-9_r8 ) then
```
Flagging this for future discussion @rgknox.  I'm wondering if we should have some sort of check or test for corrections.

glemieux commented on biogeochem/EDPatchDynamicsMod.F90:
```diff
@@ -1019,411 +1127,697 @@ subroutine set_patchno( currentSite )
-             ! above ground coarse woody debris from unburned twigs and small branches
+             ! Above ground coarse woody debris from twigs and small branches
+             ! a portion of this pool may burn
              do c = 1,2
```
Suggestion to reduce repeated lines of code:
```suggestion
             do c = 1,ncwd
                 donatable_mass = num_dead_trees * SF_val_CWD_frac(c) * bstem
                 if (c == 1 .or. c == 2) then
                      donatable_mass = donatable_mass * (1.0_r8-currentCohort%fraction_crown_burned)
                      burned_mass = num_dead_trees * SF_val_CWD_frac(c) * bstem * & 
                      currentCohort%fraction_crown_burned
                      site_mass%burn_flux_to_atm = site_mass%burn_flux_to_atm + burned_mass
                endif
                new_litt%ag_cwd(c) = new_litt%ag_cwd(c) + donatable_mass * donate_m2
                curr_litt%ag_cwd(c) = curr_litt%ag_cwd(c) + donatable_mass * retain_m2
                flux_diags%cwd_ag_input(c) = flux_diags%cwd_ag_input(c) + donatable_mass
             enddo   
                  
```

rgknox replied:
I like this.

It's hard on this thread, to tell what exactly would be removed if I commit the suggestion.  Does this include the removal of the loop here (L1582):

https://github.com/NGEET/fates/pull/524/files/3d56ddd8b5c00ac4ee98db7aedf9fa34f1097d74#diff-531e45e75218bb11634691970470c90bR1582

and it's contents?  I think that needs to be removed too.  I suppose i could just commit and see...

glemieux replied:
Yeah, sorry.  First time suggesting more than a one line change.  Yup, `the n = 3,4` section would be removed.

rgknox replied:
great, will incorporate your changes soon

rgknox commented on biogeochem/EDCohortDynamicsMod.F90:
```diff
@@ -736,52 +727,14 @@ subroutine terminate_cohorts( currentSite, currentPatch, level )
+                  currentCohort,currentCohort%n)
           end if
           
-          ! Zero out the state pools
```
Good question.  This is because this routine is meant to send the whole plant to the litter pool. The SetState() routines adjust the per-individual mass pools for each plant organ.  Since we are sending whole plants to the litter pool,  we don't need to modify the component pools of the plant, we simply need to just reduce the number density.

rgknox commented on biogeochem/EDCohortDynamicsMod.F90:
```diff
@@ -1607,7 +1676,7 @@ subroutine copy_cohort( currentCohort,copyc )
 
   !-------------------------------------------------------------------------------------!
-  function count_cohorts( currentPatch ) result ( backcount )
+  subroutine count_cohorts( currentPatch )
```
The subroutine does set a global cohort count, which is used in the photosynthesis loop.  Since this was originally a function, I had overlooked that, thinking that the function's result wasn't even used, and therefore the function isn't used.

rgknox commented on biogeochem/EDLoggingMortalityMod.F90:
```diff
@@ -1,539 +1,637 @@
-      ! This portion is known as "trunk_product_site
-
-
-      currentSite%flux_out = currentSite%flux_out + trunk_product_site
```
the flux_out variable was used for mass conservation checks.  Here is the new, equivalent line that tracks mass conservation changes from wood exporting:
https://github.com/NGEET/fates/pull/524/files/3d56ddd8b5c00ac4ee98db7aedf9fa34f1097d74#diff-135c876b8ed5286434c050b3b223605dR530-R531

rgknox replied:
We do also keep things that track diagnostics, but I wanted to create a clearer separation of variable types.  So I created the site_massbal_type which only holds variables used to ensure we are not leaking mass, and the site_fluxdiags_type that is used just for diagnostics.

https://github.com/rgknox/fates/blob/parteh-acnp-merged/main/EDTypesMod.F90#L557

The original carbon-only logging-specific diagnostics should still be preserved in this PR.

Note also, that this particular module had windows-specific formatting characters in it, on every line, which had to be removed. This makes the comparison particularly difficult to read.

rgknox commented on biogeochem/EDPhysiologyMod.F90:
```diff
@@ -898,173 +1043,244 @@ subroutine phenology_leafonoff(currentSite)
-       do while(associated(currentPatch))
+       do while (associated(currentPatch))
+          
+          litt => currentPatch%litter(el)
```
agreed, will remove that.

rgknox commented on biogeochem/EDPhysiologyMod.F90:
```diff
@@ -898,173 +1043,244 @@ subroutine phenology_leafonoff(currentSite)
+       ! equally distribute to the PFT pools
+       if ( homogenize_seed_pfts ) then
+          mean_site_seed_rain = sum(site_seed_rain(:))/real(numpft,r8)
+          site_seed_rain(1:numpft) = mean_site_seed_rain
```
yeah, I suppose that doesn't need to be two steps.  I'm pretty sure that fortran wouldn't muck  up that order of operations, ie re-calculating the sum as it goes from index to index in the new vector...right?

glemieux replied:
That would be my expectation too <fingers crossed emoji>, but that's a good question.

glemieux replied:
Looks like it should work.  I threw together a quick check: https://github.com/glemieux/fates-scratch/blob/master/fortran-scratch/summationcheck.F90.
```
$ ./summationcheck 
   2.00000000       2.00000000       0.00000000       0.00000000    
   1.00000000       2.00000000       3.00000000       4.00000000
```

rgknox commented on biogeochem/EDPhysiologyMod.F90:
```diff
@@ -1189,223 +1575,345 @@ subroutine CWD_Input( currentSite, currentPatch)
       
       do c = 1,ncwd
-         currentPatch%cwd_AG_in(c) = currentPatch%cwd_AG_in(c) + &
-              (sapw_c_turnover + struct_c_turnover)/hlm_freq_day   * &
```
Its due to a change in the units of the variable.  Previously, litter fluxes had units of per year, now they are represented in units of kg/m2/day.   This is more consistent with their use, because the flux is actually applied daily.

glemieux replied:
Nice.

rgknox commented on biogeochem/EDCohortDynamicsMod.F90:
```diff
@@ -609,18 +603,19 @@ subroutine zero_cohort(cc_p)
 
   !-------------------------------------------------------------------------------------!
-  subroutine terminate_cohorts( currentSite, currentPatch, level )
+  subroutine terminate_cohorts( currentSite, currentPatch, level , call_index)
```
yes, its hard-coded.  We could remove this, I don't have anymore need to keep this argument. It may be useful for future diagnostics though.

glemieux replied:
Gotcha, so it's mainly used for diagnostics, correct?  I think that's fine.

rgknox commented on main/FatesInterfaceMod.F90:
```diff
@@ -1131,6 +1240,39 @@ subroutine fates_history_maps
+           do icwd = 1, ncwd
+               i = i+1
+               fates_hdim_elmap_levelcwd(i)  = iel
+               fates_hdim_cwdmap_levelcwd(i) = ipft
```
yes! nice find

rgknox commented:
f10 simulations are progressing, in year 80:

/glade/scratch/rgknox/fates-clm-tests/me-litt-v3-f10_f10_mg37/run

Standard regression tests (no baseline comparison) all expected PASS:
/glade/scratch/rgknox/clmed-tests/litter-acnp-v5-C4e50a445-Fc416e641.fates.cheyenne.intel

ckoven commented on biogeochem/EDPatchDynamicsMod.F90:
```diff
@@ -1019,411 +1127,697 @@ subroutine set_patchno( currentSite )
+
+          seed_mass = num_dead * store_m * EDPftvarcon_inst%allom_frbstor_repro(pft)
+
+          ! SEED DISTRIBUTION IS BREAKING MASS CONSERVATION RIGHT NOW...
```
is this still the case and if so do we need to keep track of it with an issue?

rgknox replied:
yes, I was not able to get tight mass conservation on this, I will make an issue of this once we pull.  
Note that we do allow storage mass to convert to seed in certain grasses, it just doesn't disperse and mix at the site level.

rgknox commented:
re-running regression tests, after applying updates from @glemieux
